File: backend/app/services/email.py
import base64
import quopri
from bs4 import BeautifulSoup
from datetime import datetime, timezone


def _decode_part(data: str, transfer_encoding: str | None = None) -> str:
    try:
        padded = data + "=" * (-len(data) % 4)
        decoded = base64.urlsafe_b64decode(padded.encode("utf-8"))
        if transfer_encoding and "quoted-printable" in transfer_encoding.lower():
            decoded = quopri.decodestring(decoded)
        return decoded.decode("utf-8", errors="ignore")
    except Exception:
        return ""


def _get_transfer_encoding(part: dict) -> str | None:
    for header in part.get("headers", []) or []:
        if header.get("name", "").lower() == "content-transfer-encoding":
            return header.get("value")
    return None


async def _get_part_data(part: dict, fetch_attachment) -> str:
    body = part.get("body", {}) or {}
    encoding = _get_transfer_encoding(part)
    data = body.get("data")
    if data:
        return _decode_part(data, encoding)
    attachment_id = body.get("attachmentId")
    if attachment_id and fetch_attachment:
        try:
            attachment_data = await fetch_attachment(attachment_id)
            if attachment_data:
                return _decode_part(attachment_data, encoding)
        except Exception:
            return ""
    return ""


async def _extract_gmail_body(payload: dict, fetch_attachment) -> str:
    if not payload:
        return ""
    parts = payload.get("parts", []) or []
    if parts:
        # Prefer HTML over plain text when available.
        for part in parts:
            if part.get("mimeType") == "text/html":
                data = await _get_part_data(part, fetch_attachment)
                if data:
                    return data
        for part in parts:
            if part.get("mimeType") == "text/plain":
                data = await _get_part_data(part, fetch_attachment)
                if data:
                    return data
        for part in parts:
            nested = await _extract_gmail_body(part, fetch_attachment)
            if nested:
                return nested
        return ""

    mime = payload.get("mimeType") or ""
    if "body" in payload and (mime in {"text/html", "text/plain"} or not mime):
        data = await _get_part_data(payload, fetch_attachment)
        if data:
            return data
    return ""


async def parse_gmail_message(message: dict, fetch_attachment) -> dict:
    payload = message.get("payload", {})
    headers = payload.get("headers", []) or []
    header_map = {h["name"].lower(): h["value"] for h in headers if "name" in h}
    body = await _extract_gmail_body(payload, fetch_attachment)
    subject = header_map.get("subject")
    sender = header_map.get("from", "")
    sender_email = sender.split("<")[-1].replace(">", "").strip() if sender else ""
    sender_name = sender.split("<")[0].strip() if "<" in sender else sender
    internal_date = message.get("internalDate")
    received_at = None
    if internal_date:
        try:
            received_at = datetime.fromtimestamp(int(internal_date) / 1000, tz=timezone.utc)
        except Exception:
            received_at = None
    return {
        "message_id": message.get("id"),
        "thread_id": message.get("threadId"),
        "subject": subject,
        "sender_name": sender_name,
        "sender_email": sender_email,
        "received_at": received_at,
        "snippet": message.get("snippet"),
        "body": body,
        "headers": header_map,
    }


def parse_outlook_message(message: dict) -> dict:
    sender = message.get("from", {}).get("emailAddress", {})
    return {
        "message_id": message.get("id"),
        "thread_id": message.get("conversationId"),
        "subject": message.get("subject"),
        "sender_name": sender.get("name"),
        "sender_email": sender.get("address", ""),
        "received_at": message.get("receivedDateTime"),
        "snippet": message.get("bodyPreview"),
        "body": message.get("body", {}).get("content", ""),
        "headers": {},
    }


def clean_html(html: str) -> str:
    if not html:
        return ""
    soup = BeautifulSoup(html, "html.parser")
    for script in soup(["script"]):
        script.decompose()
    return str(soup)


def extract_preview(text: str, max_len: int = 200) -> str:
    if not text:
        return ""
    stripped = " ".join(text.split())
    return stripped[:max_len]

File: backend2/src/services/email.ts
import { Buffer } from "node:buffer";

import { load } from "cheerio";
import quotedPrintable from "quoted-printable";

export function cleanHtml(html: string): string {
  if (!html) return "";
  const $ = load(html);
  $("script").remove();
  return $.html();
}

export function extractPreview(text: string, maxLen = 200): string {
  return text ? text.replace(/\s+/g, " ").trim().slice(0, maxLen) : "";
}

function decodePart(data: string, transferEncoding?: string | null): string {
  try {
    const padded = data + "=".repeat((4 - (data.length % 4 || 4)) % 4);
    let decoded = Buffer.from(padded, "base64url");
    if (transferEncoding?.toLowerCase().includes("quoted-printable")) {
      decoded = Buffer.from(quotedPrintable.decode(decoded.toString("utf8")));
    }
    return decoded.toString("utf8");
  } catch {
    return "";
  }
}

function getTransferEncoding(part: any): string | null {
  for (const header of part?.headers ?? []) {
    if (String(header?.name ?? "").toLowerCase() === "content-transfer-encoding") {
      return String(header.value ?? "");
    }
  }
  return null;
}

async function getPartData(part: any, fetchAttachment: ((attachmentId: string) => Promise<string>) | null): Promise<string> {
  const encoding = getTransferEncoding(part);
  const data = part?.body?.data;
  if (data) return decodePart(data, encoding);
  const attachmentId = part?.body?.attachmentId;
  if (attachmentId && fetchAttachment) {
    try {
      const attachmentData = await fetchAttachment(attachmentId);
      return attachmentData ? decodePart(attachmentData, encoding) : "";
    } catch {
      return "";
    }
  }
  return "";
}

function normalizeContentId(value: string | null | undefined): string | null {
  if (!value) return null;
  return value.replace(/[<>]/g, "").trim().toLowerCase() || null;
}

function getHeader(part: any, name: string): string | null {
  for (const header of part?.headers ?? []) {
    if (String(header?.name ?? "").toLowerCase() === name.toLowerCase()) {
      return String(header.value ?? "");
    }
  }
  return null;
}

function isInlineImagePart(part: any): boolean {
  const disposition = String(getHeader(part, "content-disposition") ?? "").toLowerCase();
  const contentId = normalizeContentId(getHeader(part, "content-id"));
  return String(part?.mimeType ?? "").toLowerCase().startsWith("image/") && (Boolean(contentId) || disposition.includes("inline"));
}

function replaceCidSources(html: string, inlineAssets: Array<{ contentId: string; dataUrl: string }>): string {
  if (!html || !inlineAssets.length) return html;
  let output = html;
  for (const asset of inlineAssets) {
    const escaped = asset.contentId.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    output = output.replace(new RegExp(`cid:${escaped}`, "gi"), asset.dataUrl);
  }
  return output;
}

async function collectInlineGmailAssets(
  payload: any,
  fetchAttachment: ((attachmentId: string) => Promise<string>) | null
): Promise<Array<{ contentId: string; dataUrl: string }>> {
  if (!payload) return [];
  const assets: Array<{ contentId: string; dataUrl: string }> = [];
  const stack = [payload];
  while (stack.length) {
    const part = stack.pop();
    if (!part) continue;
    if (Array.isArray(part.parts)) {
      stack.push(...part.parts);
    }
    if (!isInlineImagePart(part)) continue;
    const contentId = normalizeContentId(getHeader(part, "content-id"));
    const bodyData = await getPartData(part, fetchAttachment);
    if (!contentId || !bodyData) continue;
    const mimeType = String(part.mimeType || "image/png");
    const base64 = Buffer.from(bodyData, "utf8").toString("base64");
    assets.push({ contentId, dataUrl: `data:${mimeType};base64,${base64}` });
  }
  return assets;
}

async function extractGmailBody(payload: any, fetchAttachment: ((attachmentId: string) => Promise<string>) | null): Promise<string> {
  if (!payload) return "";
  const parts = payload.parts ?? [];
  if (parts.length) {
    for (const mimeType of ["text/html", "text/plain"]) {
      for (const part of parts) {
        if (part.mimeType === mimeType) {
          const data = await getPartData(part, fetchAttachment);
          if (data) return data;
        }
      }
    }
    for (const part of parts) {
      const nested = await extractGmailBody(part, fetchAttachment);
      if (nested) return nested;
    }
    return "";
  }
  const mimeType = payload.mimeType ?? "";
  if (payload.body && (mimeType === "text/html" || mimeType === "text/plain" || !mimeType)) {
    return getPartData(payload, fetchAttachment);
  }
  return "";
}

export async function parseGmailMessage(
  message: any,
  fetchAttachment: ((attachmentId: string) => Promise<string>) | null
): Promise<Record<string, unknown>> {
  const payload = message.payload ?? {};
  const headerMap = Object.fromEntries((payload.headers ?? []).filter((h: any) => h?.name).map((h: any) => [String(h.name).toLowerCase(), h.value]));
  const body = await extractGmailBody(payload, fetchAttachment);
  const inlineAssets = await collectInlineGmailAssets(payload, fetchAttachment);
  const sender = String(headerMap.from ?? "");
  return {
    message_id: message.id,
    thread_id: message.threadId,
    subject: headerMap.subject ?? null,
    sender_name: sender.includes("<") ? sender.split("<", 1)[0].trim() : sender,
    sender_email: sender.includes("<") ? sender.split("<").at(-1)?.replace(">", "").trim() : sender,
    received_at: message.internalDate ? new Date(Number(message.internalDate)).toISOString() : null,
    snippet: message.snippet,
    body: replaceCidSources(body, inlineAssets),
    headers: headerMap
  };
}

export function parseOutlookMessage(message: any): Record<string, unknown> {
  const sender = message?.from?.emailAddress ?? {};
  const attachments: any[] = Array.isArray(message?.attachments) ? message.attachments : [];
  const inlineAssets = attachments
    .filter((attachment: any) => attachment?.isInline && attachment?.contentId && attachment?.contentBytes)
    .map((attachment: any) => ({
      contentId: normalizeContentId(String(attachment.contentId))!,
      dataUrl: `data:${String(attachment.contentType || "image/png")};base64,${String(attachment.contentBytes)}`
    }))
    .filter((asset: { contentId: string | null }) => asset.contentId);
  return {
    message_id: message.id,
    thread_id: message.conversationId,
    subject: message.subject,
    sender_name: sender.name ?? null,
    sender_email: sender.address ?? "",
    received_at: message.receivedDateTime ?? null,
    snippet: message.bodyPreview ?? null,
    body: replaceCidSources(message?.body?.content ?? "", inlineAssets),
    headers: {}
  };
}
